docs/proposals/protobuf.md

```diff
@@ -0,0 +1,501 @@
+identify the boundaries between Events (a `frame`). We do that by prefixing
+each frame of N bytes with a 4-byte, big-endian, unsigned integer with the
+value N.
+
+    frame  = length body
+    length = 32-bit unsigned integer in big-endian order, denoting length of
+             bytes of body
+    body = <bytes>
+
+    # frame containing a single byte 0a
+    frame = 01 00 00 00 0a
+
+    # equivalent JSON
+    frame = {"type": "added", ...}
+
+The body of each frame is a serialized Protobuf message `Event` in package
+`k8s.io/kubernetes/pkg/watch/versioned`. The content type used for this
+format is `application/vnd.kubernetes.protobuf;type=watch`.
+
+## Negotiation
+
+To allow clients to request protobuf serialization optionally, the `Accept`
+HTTP header is used by callers to indicate which serialization they wish
+returned in the response, and the `Content-Type` header is used to tell the
+server how to decode the bytes sent in the request (for DELETE/POST/PUT/PATCH
+requests). The server will return 406 if the `Accept` header is not
+recognized or 415 if the `Content-Type` is not recognized (as defined in
+RFC2616).
+
+To be backwards compatible, clients must consider that the server does not
+support protobuf serialization. A number of options are possible:
+
+### Preconfigured
+
+Clients can have a configuration setting that instructs them which version
+to use. This is the simplest option, but requires intervention when the
+component upgrades to protobuf.
+
+### Include serialization information in api-discovery
+
+Servers can define the list of content types they accept and return in
+their API discovery docs, and clients can use protobuf if they support it.
+Allows dynamic configuration during upgrade if the client is already using
+API-discovery.
+
+### Optimistically attempt to send and receive requests using protobuf
+
+Using multiple `Accept` values:
+
+    Accept: application/vnd.kubernetes.protobuf, application/json
+
+clients can indicate their preferences and handle the returned
+`Content-Type` using whatever the server responds. On update operations,
+clients can try protobuf and if they receive a 415 error, record that and
+fall back to JSON. Allows the client to be backwards compatible with
+any server, but comes at the cost of some implementation complexity.
+
+
+## Generation process
+
+Generation proceeds in five phases:
+
+1. Generate a gogo-protobuf annotated IDL from the source Go struct.
+2. Generate temporary Go structs from the IDL using gogo-protobuf.
+3. Generate marshaller/unmarshallers based on the IDL using gogo-protobuf.
+4. Take all tag numbers generated for the IDL and apply them as struct tags
+   to the original Go types.
+5. Generate a final IDL without gogo-protobuf annotations as the canonical IDL.
+
+The output is a `generated.proto` file in each package containing a standard
+proto2 IDL, and a `generated.pb.go` file in each package that contains the
+generated marshal/unmarshallers.
```

Comment:
I think there is value in having the generated proto checked in somewhere.
They are stable to input which means diffs should be small.  If we don't
check them in then we have no way of knowing if the generator is broken and
our proto IDL regressed between two commits without adding an additional
process that detects incompatible schemas.

On Mon, Mar 7, 2016 at 4:25 AM, Wojciech Tyczynski <notifications@github.com
> wrote:

> In docs/proposals/protobuf.md
> <https://github.com/kubernetes/kubernetes/pull/22600#discussion_r55184849>
> :
>
> > +
> > +
> > +## Generation process
> > +
> > +Generation proceeds in five phases:
> > +
> > +1. Generate a gogo-protobuf annotated IDL from the source Go struct.
> > +2. Generate temporary Go structs from the IDL using gogo-protobuf.
> > +3. Generate marshaller/unmarshallers based on the IDL using gogo-protobuf.
> > +4. Take all tag numbers generated for the IDL and apply them as struct tags
> > +   to the original Go types.
> > +5. Generate a final IDL without gogo-protobuf annotations as the canonical IDL.
> > +
> > +The output is a `generated.proto` file in each package containing a standard
> > +proto2 IDL, and a `generated.pb.go` file in each package that contains the
> > +generated marshal/unmarshallers.
>
> Are we going to commit those files to repository?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/22600/files#r55184849>.
>